plugins/qcheck-stm/src/ir_of_gospel.ml

```diff
@@ -1,49 +1,97 @@
 open Gospel
 open Tast
-open Reserr
-open Config
 
-(* XXX TODO should check the compatibility of the argument type too *)
-let lb_arg_is_of_type ts lb =
-  match lb with
-  | Lunit -> false
-  | Lnone vs | Loptional vs | Lnamed vs | Lghost vs -> (
-      match vs.vs_ty.ty_node with
-      | Tyapp (ts', _) -> Ttypes.ts_equal ts ts'
-      | _ -> false)
+let type_is_sut_type config ty =
+  let sut_type_name = Config.get_sut_type_name config in
+  let open Ppxlib in
+  match ty.ptyp_desc with
+  | Ptyp_constr (lid, _) -> lid.txt = sut_type_name
+  | _ -> false
 
-let lb_arg_is_not_of_type ty lb_arg = not (lb_arg_is_of_type ty lb_arg)
+let type_is_not_sut_type config ty = not (type_is_sut_type config ty)
 
 let constant_test vd =
+  let open Reserr in
   match vd.vd_args with
   | [] -> (Constant_value vd.vd_name.id_str, vd.vd_loc) |> error
-  | _ -> ok vd
-
-let argument_test ty vd =
-  let p = lb_arg_is_of_type ty in
-  let rec exactly_one = function
-    | [] -> (No_sut_argument vd.vd_name.id_str, vd.vd_loc) |> error
-    | x :: xs when p x ->
-        if List.exists p xs then
-          (Multiple_sut_arguments vd.vd_name.id_str, vd.vd_loc) |> error
-        else ok vd
-    | _ :: xs -> exactly_one xs
-  in
-  exactly_one vd.vd_args
+  | _ -> ok ()
+
+let is_a_function ty =
+  let open Ppxlib in
+  match ty.ptyp_desc with Ptyp_arrow (_, _, _) -> true | _ -> false
 
-let return_test ty vd =
-  if List.for_all (lb_arg_is_not_of_type ty) vd.vd_ret then ok vd
-  else (Returning_sut vd.vd_name.id_str, vd.vd_loc) |> error
+let unify value_name sut_ty ty =
+  let open Ppxlib in
+  let open Reserr in
+  let add_if_needed a x i =
+    match List.assoc_opt a i with
+    | None -> ok ((a, x) :: i)
+    | Some y when x.ptyp_desc = y.ptyp_desc -> ok i
+    | _ -> error (Incompatible_type value_name, ty.ptyp_loc)
+  in
+  let rec aux i = function
+    | [], [] -> ok i
+    | x :: xs, y :: ys ->
+        let* i =
+          match (x.ptyp_desc, y.ptyp_desc) with
+          | _, Ptyp_var a -> add_if_needed a x i
+          | Ptyp_tuple xs, Ptyp_tuple ys -> aux i (xs, ys)
+          | Ptyp_constr (c, xs), Ptyp_constr (d, ys) when c.txt = d.txt ->
+              aux i (xs, ys)
+          | _ -> error (Incompatible_type value_name, ty.ptyp_loc)
+        in
+        aux i (xs, ys)
+    | _, _ -> error (Incompatible_type value_name, ty.ptyp_loc)
+  in
+  match (sut_ty.ptyp_desc, ty.ptyp_desc) with
+  | Ptyp_constr (t, args_sut), Ptyp_constr (t', args_ty) when t.txt = t'.txt ->
+      aux [] (args_sut, args_ty)
+  | _ -> invalid_arg "Impossible unification."
```

Comment:
This is using `invalid_arg` because we are in a “should never happen” case, aren’t we?

Reply:
Yes, this case should never happen:
- we've only accepted `Ptyp_constr` as type for sut (at least for now, we need to remember that if we change that)
- we check that `sut_ty` and `ty` have the same type constructor before calling `unify`.

I'll change to use `failwith` which makes more sense.